src/rgw/rgw_asio_frontend.cc

```diff
@@ -65,67 +62,103 @@ void Pauser::wait()
-  boost::system::error_code ec;
+class Connection {
+  RGWProcessEnv& env;
+  boost::asio::strand strand;
+  tcp::socket socket;
 
-  // limit header to 4k, since we read it all into a single buffer
-  constexpr size_t header_limit = 4096;
+  // references are bound to callbacks for async operations. if a callback
+  // function returns without issuing another operation, the reference is
+  // dropped and the Connection is deleted/closed
+  std::atomic<int> nref{0};
+  using Ref = boost::intrusive_ptr<Connection>;
+
+  // limit header to 4k, since we read it all into a single flat_buffer
+  static constexpr size_t header_limit = 4096;
   // don't impose a limit on the body, since we read it in pieces
-  constexpr size_t body_limit = std::numeric_limits<size_t>::max();
+  static constexpr size_t body_limit = std::numeric_limits<size_t>::max();
 
   beast::flat_buffer buffer{header_limit};
+  boost::optional<rgw::asio::parser_type> parser;
+
+  using bad_response_type = beast::http::response<beast::http::empty_body>;
+  boost::optional<bad_response_type> response;
+
+  CephContext* ctx() const { return env.store->ctx(); }
 
-  // read messages from the socket until eof
-  for (;;) {
+  void read_header() {
     // configure the parser
-    rgw::asio::parser_type parser;
-    parser.header_limit(header_limit);
-    parser.body_limit(body_limit);
+    parser.emplace();
+    parser->header_limit(header_limit);
+    parser->body_limit(body_limit);
 
     // parse the header
-    beast::http::async_read_header(socket, buffer, parser, yield[ec]);
+    beast::http::async_read_header(socket, buffer, *parser, strand.wrap(
+            std::bind(&Connection::on_header, Ref{this},
+                      std::placeholders::_1)));
+  }
+
+  void on_write_error(boost::system::error_code ec) {
+    if (ec) {
+      ldout(ctx(), 5) << "failed to write response: " << ec.message() << dendl;
+    }
+  }
 
+  void on_header(boost::system::error_code ec) {
     if (ec == boost::asio::error::connection_reset ||
-        ec == boost::asio::error::eof) {
+        ec == beast::http::error::end_of_stream) {
       return;
     }
     if (ec) {
-      auto& message = parser.get();
-      ldout(cct, 1) << "read failed: " << ec.message() << dendl;
-      ldout(cct, 1) << "====== req done http_status=400 ======" << dendl;
-      beast::http::response<beast::http::string_body> response;
-      response.result(beast::http::status::bad_request);
-      response.reason("Bad Request");
-      response.version(message.version() == 10 ? 10 : 11);
-      response.prepare_payload();
-      beast::http::async_write(socket, response, yield[ec]);
-      // ignore ec
+      auto& message = parser->get();
+      ldout(ctx(), 1) << "failed to read header: " << ec.message() << dendl;
+      ldout(ctx(), 1) << "====== req done http_status=400 ======" << dendl;
+      response.emplace();
+      response->result(beast::http::status::bad_request);
+      response->reason("Bad Request");
```

Comment:
If you don't set the reason explicitly, Beast will fill in an appropriate string for you during serialization